Sources/PackagePlugin/Implementation.swift

```diff
@@ -0,0 +1,132 @@
+            let decoder = JSONDecoder()
+            context = try decoder.decode(TargetBuildContext.self, from: inputData)
+        }
+        catch {
+            fputs("Couldn't decode input JSON (reason: \(error)", stderr)
+            output.diagnostics.append(Diagnostic(severity: .error, message: "\(error)", file: nil, line: nil))
+            exit(1)
+        }
+
+        // Instantiate the plugin. For now there are no parameters, but this is where we would set them up, most likely as properties of the plugin.
+        let plugin = self.init()
+        
+        // Invoke the appropriate protocol method based on the action.
+        switch context.pluginAction {
+        case .createBuildToolCommands:
+            // Check that the plugin conforms to `BuildToolPlugin`, and get the commands.
+            guard let plugin = plugin as? BuildToolPlugin else { throw PluginDeserializationError.malformedInputJSON("...") }
+            let commands = try plugin.createBuildCommands(context: context)
+            
+            // Convert the commands to the encodable output representation SwiftPM currently expects.
+            output.buildCommands = commands.compactMap {
+                guard case let ._buildCommand(displayName, executable, arguments, environment, workingDir, inputFiles, outputFiles) = $0 else { return .none }
+                return BuildCommand(displayName: displayName, executable: executable, arguments: arguments, environment: environment, workingDirectory: workingDir, inputFiles: inputFiles, outputFiles: outputFiles)
+            }
+            output.prebuildCommands = commands.compactMap {
+                guard case let ._prebuildCommand(displayName, executable, arguments, environment, workingDir, outputFilesDir) = $0 else { return .none }
+                return PrebuildCommand(displayName: displayName, executable: executable, arguments: arguments, environment: environment, workingDirectory: workingDir, outputFilesDirectory: outputFilesDir)
+            }
+        }
+
+        // Emit the output from the plugin for SwiftPM to read.
+        let encoder = JSONEncoder()
+        let outputData = try! encoder.encode(output)
+        fputc(0, stdout)
+        fputs(String(data: outputData, encoding: .utf8)!, stdout)
+        fflush(stdout)
+    }
+}
+
+/// Private structures containing the information to send back to SwiftPM.
+
+struct BuildCommand: Encodable {
+    let displayName: String?
+    let executable: Path
+    let arguments: [String]
+    let environment: [String: String]
+    let workingDirectory: Path?
+    let inputFiles: [Path]
+    let outputFiles: [Path]
+}
+
+struct PrebuildCommand: Encodable {
+    let displayName: String?
+    let executable: Path
+    let arguments: [String]
+    let environment: [String: String]
+    let workingDirectory: Path?
+    let outputFilesDirectory: Path
+}
+
+struct Diagnostic: Encodable {
+    enum Severity: String, Encodable {
+        case error, warning, remark
```

Comment:
remark -> info ?

Reply:
Maybe.  Or `note`.  In the amendment proposal this is `remark`, so we should also take it up there whether it should be something else.